clippy_lints/src/doc/doc_comments_missing_terminal_punctuation.rs

```diff
@@ -0,0 +1,88 @@
+use rustc_errors::Applicability;
+use rustc_lint::LateContext;
+use rustc_resolve::rustdoc::main_body_opts;
+
+use pulldown_cmark::{Event, Options, Parser, Tag, TagEnd};
+
+use super::{DOC_COMMENTS_MISSING_TERMINAL_PUNCTUATION, Fragments};
+
+const MSG: &str = "doc comments should end with a terminal punctuation mark";
+const PUNCTUATION_SUGGESTION: char = '.';
+
+pub fn check(cx: &LateContext<'_>, doc: &str, fragments: Fragments<'_>) {
+    // This ignores `#[doc]` attributes, which we do not handle.
+    if let Some(offset) = is_missing_punctuation(doc)
+        && let Some(span) = fragments.span(cx, offset..offset)
+    {
+        clippy_utils::diagnostics::span_lint_and_sugg(
+            cx,
+            DOC_COMMENTS_MISSING_TERMINAL_PUNCTUATION,
+            span,
+            MSG,
+            "end the doc comment with some punctuation",
+            PUNCTUATION_SUGGESTION.to_string(),
+            Applicability::MaybeIncorrect,
+        );
+    }
+}
+
+#[must_use]
+/// If punctuation is missing, returns the offset where new punctuation should be inserted.
+fn is_missing_punctuation(doc_string: &str) -> Option<usize> {
+    const TERMINAL_PUNCTUATION_MARKS: &[char] = &['.', '?', '!', '…'];
+
+    let mut no_report_depth = 0;
+    let mut text_offset = None;
+    for (event, offset) in
+        Parser::new_ext(doc_string, main_body_opts() - Options::ENABLE_SMART_PUNCTUATION).into_offset_iter()
+    {
+        match event {
+            Event::Start(
+                Tag::CodeBlock(..)
+                | Tag::FootnoteDefinition(_)
+                | Tag::Heading { .. }
+                | Tag::HtmlBlock
+                | Tag::List(..)
+                | Tag::Table(_),
+            ) => {
+                no_report_depth += 1;
+            },
+            Event::End(TagEnd::FootnoteDefinition) => {
+                no_report_depth -= 1;
+            },
+            Event::End(
+                TagEnd::CodeBlock | TagEnd::Heading(_) | TagEnd::HtmlBlock | TagEnd::List(_) | TagEnd::Table,
+            ) => {
+                no_report_depth -= 1;
+                text_offset = None;
+            },
+            Event::InlineHtml(_) | Event::Start(Tag::Image { .. }) | Event::End(TagEnd::Image) => {
+                text_offset = None;
+            },
+            Event::Code(..) | Event::Start(Tag::Link { .. }) | Event::End(TagEnd::Link)
+                if no_report_depth == 0 && !offset.is_empty() =>
+            {
+                text_offset = Some(offset.end);
+            },
+            Event::Text(..) if no_report_depth == 0 && !offset.is_empty() => {
+                // American-style quotes require punctuation to be placed inside closing quotation marks.
+                if doc_string[..offset.end].trim_end().ends_with('"') {
```

Comment:
Okay, so this code can be extended to handle parens with basically-trivial additional complexity, right?


```suggestion
                if doc_string[..offset.end].trim_end().ends_with(|c| c == '"' || c == ')') {
```

Reply:
> Maybe we should treat these cases (quotes and parentheses) as unfixable: detect when terminal punctuation is missing but give no suggestions?

Yes, I agree. That would be great!

Reply:
> To be honest I think we are overfocusing on edge cases right now: this lint (even without handling quotes and parentheses) would already be really useful for rolling out on new projects to ensure a consistent style, and we know that we will not be able to handle every little edge case without full-blown NLP anyway.

It's certainly true that we can't reach perfection without full AI. When we inevitably err, though, I'd prefer to err on the side of false negatives instead of false positives. We could even solve the parens and quotes problem by adding them as accepted terminating punctuation:

```rust
const TERMINAL_PUNCTUATION_MARKS: &[char] = &['.', '?', '!', '…', ')', '"', '\''];
```

The point is that we don't want to make people `#[allow]` the lint.